dataplane/linux/int_dataplane.go

```diff
@@ -301,28 +321,39 @@ func NewIntDataplaneDriver(config Config) *InternalDataplane {
 		config.RulesConfig.IptablesMarkNonCaliEndpoint)
 
 	// Auto-detect host MTU.
-	if mtu, err := findHostMTU(config.MTUIfacePattern); err != nil {
+	hostMTU, err := findHostMTU(config.MTUIfacePattern)
+	if err != nil {
 		log.WithError(err).Fatal("Unable to detect host MTU, shutting down")
-	} else {
-		// We found the host's MTU. Default any MTU configurations that have not been set.
-		// We default the values even if the encap is not enabled, in order to match behavior
-		// from earlier versions of Calico. However, they MTU will only be considered for allocation
-		// to pod interfaces if the encap is enabled.
-		config.hostMTU = mtu
-		if config.IPIPMTU == 0 {
-			log.Debug("Defaulting IPIP MTU based on host")
-			config.IPIPMTU = mtu - 20
-		}
-		if config.VXLANMTU == 0 {
-			log.Debug("Defaulting VXLAN MTU based on host")
-			config.VXLANMTU = mtu - 50
-		}
-		if config.Wireguard.MTU == 0 {
-			log.Debug("Defaulting Wireguard MTU based on host")
-			config.Wireguard.MTU = mtu - 60
+	}
+	// We found the host's MTU. Default any MTU configurations that have not been set.
+	// We default the values even if the encap is not enabled, in order to match behavior
+	// from earlier versions of Calico. However, they MTU will only be considered for allocation
+	// to pod interfaces if the encap is enabled.
+	config.hostMTU = hostMTU
+	if config.IPIPMTU == 0 {
+		log.Debug("Defaulting IPIP MTU based on host")
+		config.IPIPMTU = hostMTU - ipipMTUOverhead
+	}
+	if config.VXLANMTU == 0 {
+		log.Debug("Defaulting VXLAN MTU based on host")
+		config.VXLANMTU = hostMTU - vxlanMTUOverhead
+	}
+	if config.Wireguard.MTU == 0 {
+		// The default MTU on Azure is 1500, but the underlying network stack will fragment packets at 1400 bytes, see
+		// https://docs.microsoft.com/en-us/azure/virtual-network/virtual-network-tcpip-performance-tuning#azure-and-vm-mtu
+		// for details.
+		// Additionally, Wireguard sets the DF bit on its packets, and so if the MTU is set too high large packets will
+		// be dropped.
+		// Therefore it is necessary to allow for the difference between the MTU of the host and the underlying network.
+		if config.KubernetesProvider == ProviderAKS && config.RouteSource == "WorkloadIPs" {
+			log.Debug("Defaulting Wireguard MTU based on host and AKS with WorkloadIPs")
+			config.Wireguard.MTU = hostMTU - aksMTUOverhead - wireguardMTUOverhead
 		}
+		log.Debug("Defaulting Wireguard MTU based on host")
+		config.Wireguard.MTU = hostMTU - wireguardMTUOverhead
 	}
-	if err := writeMTUFile(config); err != nil {
+	podMTU := determinePodMTU(config)
+	if err := writeMTUFile(podMTU); err != nil {
```

Comment:
I'm not sure on the best way to unit test the MTU calculation above. Since the `config` param is not a pointer, the test can't inspect the changes made to the config, so maybe abstract the writing of the MTU file here so I can test the `podMTU` it gets called with?

Reply:
Why not extract out a new function that takes a pointer `detectMTUs(config *Config)`?  (This function is too big anyway!)

Reply:
That looks better, but to be able to call that new function from the `intdataplane_test` package it'll need to be public, is that ok?

Reply:
You can do either in Go, this package has examples of both.  E.g. endpoint_mgr_test.go is in the main package.  Feel free to move int_dataplane_test.go to the non-test package if that makes it easier to test.

To expand on my thinking (since it's more nuanced than "anything goes :-P": I think the idea behind the `_test` package is several-fold:

* it allows for breaking certain kinds of import cycles (not sure I grok this but I read it somewhere!)
* it encourages blackbox testing where you test the function rather than the internals
* it lets you test from the user's point of view; verify that you actually exposed the API you were expecting.

All good things but 
* This isn't an "API" package so I'm not worried about the third point
* MTU detection feels like a nice granularity of "unit" to test (so it feels like it meets the second bullet in spirit, whether you test from the same package or `_test`).  An alternative would be to move MTU detection logic out into its own package; might be a nice cleanup but also not particularly pressing IMO.

Reply:
Thanks for detailed explanation! I tried passing in a `MTUFileWriter` which I could then have a fake implementation of in the tests, and it worked fine, but looked a little complicated. In the end I just did as you first suggested and extracted a function which takes a pointer to the config.